lib/core/test/unit/Cardano/Pool/MetadataSpec.hs

```diff
@@ -0,0 +1,239 @@
+    , mkClient
+    , newManager
+    )
+import Cardano.Wallet.Api.Server
+    ( Listen (..), withListeningSocket )
+import Cardano.Wallet.Api.Types
+    ( ApiT (..) )
+import Cardano.Wallet.Primitive.Types
+    ( PoolId (..), StakePoolOffChainMetadata (..), StakePoolTicker (..) )
+import Control.Concurrent.Async
+    ( async, cancel )
+import Control.Concurrent.MVar
+    ( newEmptyMVar, putMVar, takeMVar )
+import Control.Exception
+    ( bracket )
+import Control.Lens
+    ( at, (.~), (?~) )
+import Data.Aeson
+    ( toJSON )
+import Data.Function
+    ( (&) )
+import Data.Proxy
+    ( Proxy (..) )
+import Data.String
+    ( fromString )
+import Data.Swagger
+    ( Referenced (..)
+    , Schema
+    , SwaggerType (..)
+    , maxLength
+    , minLength
+    , paramSchema
+    , properties
+    , required
+    , type_
+    )
+import Data.Swagger.Schema.Validation
+    ( validateJSON )
+import Data.Text
+    ( Text )
+import Data.Text.Class
+    ( ToText (..) )
+import Data.Vector.Shuffle
+    ( mkSeed )
+import Network.HTTP.Types
+    ( status404 )
+import Network.Wai.Handler.Warp
+    ( runSettingsSocket, setBeforeMainLoop )
+import Servant
+    ( Server, err404, serve, throwError )
+import Servant.Client
+    ( ClientError (..), responseStatusCode )
+import Test.Hspec
+    ( Spec, around, describe, it )
+import Test.QuickCheck
+    ( ASCIIString (..)
+    , Arbitrary (..)
+    , PrintableString (..)
+    , choose
+    , counterexample
+    , label
+    , property
+    , vector
+    , vectorOf
+    )
+import Test.QuickCheck.Gen
+    ( Gen (..) )
+import Test.QuickCheck.Monadic
+    ( assert, monadicIO, monitor, run )
+import Test.QuickCheck.Random
+    ( mkQCGen )
+
+import qualified Data.Aeson as Aeson
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Lazy.Char8 as BL8
+import qualified Data.Text as T
+import qualified Network.Wai.Handler.Warp as Warp
+
+spec :: Spec
+spec = describe "Metadata - MockServer" $
+    around withMockServer $ it "Mock Server works as intended" $
+        \Client{getStakePoolMetadata} -> property $ \pid -> monadicIO $ do
+            run (getStakePoolMetadata pid) >>= \case
+                Left (FailureResponse _ res) -> do
+                    monitor $ label "No Corresponding Metadata"
+                    monitor $ counterexample $ mconcat
+                        [ show (responseStatusCode res)
+                        , " =/= "
+                        , show status404
+                        ]
+                    assert (responseStatusCode res == status404)
+                Left e -> do
+                    monitor $ counterexample $
+                        "unexpected response from server: " <> show e
+                    assert False
+                Right !metadata -> do
+                    monitor $ label "Got Valid Metadata"
+                    let json = toJSON $ ApiT metadata
+                    let errs = validateJSON mempty metadataSchema json
```

Comment:
👌 

Would it be interesting to see how the client reacts when the response does _not_ conform to the swagger spec?

Reply:
There's no logic in the client so far; I have another chunk of work adding the interesting logic of fetching / caching the metadata. Here it'll be more interesting to test faulty behavior. 

Also, the server not conforming to the swagger schema will likely result in the JSON parser failing, and therefore, a `Left ClientError` on the client. The JSON parser is quite lenient however, it doesn't check for things like minimal or maximal length. 

Having said that, the server not satisfying the interface is a "contract violation" and something that would require a fix on the server, if / when possible.